File: server/shell_backend.ts
import type { ShellRequest, ShellResponse } from "../type/rpc.ts";
import type { ServerOptions } from "./http_server.ts";

/**
 * Configuration via environment variables:
 * - SB_SHELL_BACKEND: "local" or "off"
 */
export function determineShellBackend(
  serverOptions: ServerOptions,
): ShellBackend {
  const backendConfig = Deno.env.get("SB_SHELL_BACKEND") || "local";
  switch (backendConfig) {
    case "local":
      if (serverOptions.shellCommandWhiteList) {
        console.info(
          "Running with the following shell commands enabled:",
          serverOptions.shellCommandWhiteList,
        );
      } else {
        console.info("Running with ALL shell commands enabled.");
      }
      return new LocalShell(
        serverOptions.pagesPath,
        serverOptions.shellCommandWhiteList,
      );
    default:
      console.info(
        "Running in shell-less mode, meaning shell commands are disabled",
      );
      return new NotSupportedShell();
  }
}

export interface ShellBackend {
  handle(shellRequest: ShellRequest): Promise<ShellResponse>;
}

export class NotSupportedShell implements ShellBackend {
  handle(): Promise<ShellResponse> {
    return Promise.resolve({
      stdout: "",
      stderr: "Not supported",
      code: 1,
    });
  }
}

export class LocalShell implements ShellBackend {
  constructor(private cwd: string, private cmdWhiteList?: string[]) {
  }

  async handle(shellRequest: ShellRequest): Promise<ShellResponse> {
    if (this.cmdWhiteList && !this.cmdWhiteList.includes(shellRequest.cmd)) {
      console.error(
        "Not running shell command because not in whitelist",
        shellRequest,
      );
      return {
        code: -1,
        stdout: "",
        stderr: "Not allowed, command not in whitelist",
      };
    }
    console.log(
      "Running shell command:",
      shellRequest.cmd,
      shellRequest.args,
    );
    const p = new Deno.Command(shellRequest.cmd, {
      cwd: this.cwd,
      args: shellRequest.args,
      stdin: "piped",
      stdout: "piped",
      stderr: "piped",
    }).spawn();
    if (shellRequest.stdin) {
      // Write a string to stdin
      const encoder = new TextEncoder();
      const writer = p.stdin.getWriter();
      await writer.write(encoder.encode(shellRequest.stdin));
      writer.releaseLock();
    }
    await p.stdin.close();

    // Capture stdout and stderr
    const [stdout, stderr] = await Promise.all([
      readAll(p.stdout.getReader()),
      readAll(p.stderr.getReader()),
    ]);

    // Get the exit status
    const status = await p.status;

    return {
      stderr,
      stdout,
      code: status.code,
    };
  }
}

async function readAll(r: ReadableStreamDefaultReader) {
  const chunks: string[] = [];
  const decoder = new TextDecoder();
  while (true) {
    const { value, done } = await r.read();
    if (done) break;
    chunks.push(decoder.decode(value));
  }
  return chunks.join();
}

File: server/shell_backend.go
package server

import (
	"bytes"
	"log"
	"os/exec"
	"slices"
	"strings"
)

// LocalShell implements shell execution on the local system
type LocalShell struct {
	Cwd          string
	CmdWhiteList []string
	AllowAllCmds bool
}

// NewLocalShell creates a new LocalShell instance
func NewLocalShell(cwd string, cmdWhiteList string) *LocalShell {
	var whiteListedCommands []string
	if cmdWhiteList == "" {
		whiteListedCommands = []string{}
	} else {
		// Use comma as delimiter to support paths with spaces
		parts := strings.Split(cmdWhiteList, ",")
		for _, part := range parts {
			trimmed := strings.TrimSpace(part)
			if trimmed != "" {
				whiteListedCommands = append(whiteListedCommands, trimmed)
			}
		}
	}
	return &LocalShell{
		Cwd:          cwd,
		CmdWhiteList: whiteListedCommands,
		AllowAllCmds: len(whiteListedCommands) == 0,
	}
}

// Handle executes a shell command and returns the result
func (ls *LocalShell) Handle(request ShellRequest) (ShellResponse, error) {
	// Check if command is whitelisted
	if !ls.AllowAllCmds && !slices.Contains(ls.CmdWhiteList, request.Cmd) {
		log.Printf("Not running shell command because not in whitelist: %s\n", request.Cmd)
		return ShellResponse{
			Code:   -1,
			Stdout: "",
			Stderr: "Not allowed, command not in whitelist",
		}, nil
	}

	log.Printf("Running shell command: %s %s\n", request.Cmd, strings.Join(request.Args, " "))

	// Create the command
	cmd := exec.Command(request.Cmd, request.Args...)
	cmd.Dir = ls.Cwd

	// Set up stdin if provided
	if request.Stdin != nil {
		cmd.Stdin = strings.NewReader(*request.Stdin)
	}

	// Set up separate buffers for stdout and stderr
	var stdout, stderr bytes.Buffer
	cmd.Stdout = &stdout
	cmd.Stderr = &stderr

	// Execute the command
	err := cmd.Run()

	var exitCode int
	if err != nil {
		// Try to get the exit code
		if exitError, ok := err.(*exec.ExitError); ok {
			exitCode = exitError.ExitCode()
		} else {
			// Command failed to start
			return ShellResponse{
				Code:   -1,
				Stdout: "",
				Stderr: err.Error(),
			}, nil
		}
	}

	return ShellResponse{
		Code:   exitCode,
		Stdout: stdout.String(),
		Stderr: stderr.String(),
	}, nil
}

// NotSupportedShell implements a shell backend that doesn't support execution
type NotSupportedShell struct{}

// NewNotSupportedShell creates a new NotSupportedShell instance
func NewNotSupportedShell() *NotSupportedShell {
	return &NotSupportedShell{}
}

// Handle always returns a "not supported" error
func (nss *NotSupportedShell) Handle(request ShellRequest) (ShellResponse, error) {
	return ShellResponse{
		Code:   1,
		Stdout: "",
		Stderr: "Not supported",
	}, nil
}
